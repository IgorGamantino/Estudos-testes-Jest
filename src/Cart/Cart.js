import find, { remove } from 'lodash'
import Dinero from 'dinero.js'


Dinero.defaultCurrency = 'BRL';
Dinero.defaultPrecision = 2;


export class Cart {
  cart = []

  add(products) {
    if (find(this.cart, { product: products.product })) {
      remove(this.cart, { product: products.product })
    }
    this.cart.push(products)
  }

  remove(products) {
    remove(this.cart, { products })
  }

  getTotal() {
    return this.cart.reduce((acc, item) => {
      const amount = Dinero({ amount: item.quantity * item.product.price })
      let discount = Dinero({ amount: 0 })


      if (item.condition && item.condition.percentage && item.quantity > 2) {
        discount = amount.percentage(item.condition.percentage
        )
      }

      return acc.add(amount).subtract(discount)
    }, Dinero({ amount: 0 }))
  }

  summary() {
    const total = this.getTotal().getAmount();
    const items = this.cart;

    return {
      total,
      items
    }
  }

  checkout() {
    const total = this.getTotal()
    const items = this.cart;

    this.cart = []

    return {
      total,
      items
    }
  }

}