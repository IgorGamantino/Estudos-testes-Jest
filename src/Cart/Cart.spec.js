import { Cart } from './Cart'

describe('Cart', () => {
  let cart;
  let product = {
    title: 'tenis adidas',
    price: 35

  }
  let productTwo = {
    title: 'tenis nike',
    price: 80

  }

  beforeEach(() => {
    cart = new Cart()
  })

  describe('getTotal()', () => {

    it('shoud be return 0 when getTotal() is executed in a newly created instance', () => {

      expect(cart.getTotal()).toEqual(0)
    })

    it('shoud be multiply quantity and price and receive the total amount', () => {
      const item = {
        product,
        quantity: 2
      }
      cart.add(item)

      expect(cart.getTotal()).toEqual(70)

    })

    it('shoud be ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 2
      })

      cart.add({
        product,
        quantity: 1,
      })

      expect(cart.getTotal()).toEqual(35)
    })

    it('shoud be update total when a product gets included and then remove', () => {
      cart.add({
        product,
        quantity: 2
      })

      cart.add({
        product: productTwo,
        quantity: 1
      })

      cart.remove(
        product,

      )

      expect(cart.getTotal()).toEqual(150)

    })
  })

  describe('checkout', () => {
    it('should be return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 2
      })

      cart.add({
        product: productTwo,
        quantity: 6
      })

      expect(cart.checkout()).toMatchSnapshot()

    })

    it('should be reset the cart when checkout() is called', () => {
      cart.add({
        product,
        quantity: 3
      })

      cart.checkout()


      expect(cart.getTotal()).toEqual(0)

    })
  })

})

