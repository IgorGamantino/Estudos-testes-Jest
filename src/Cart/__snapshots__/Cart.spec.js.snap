// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout should be return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35,
        "title": "tenis adidas",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 80,
        "title": "tenis nike",
      },
      "quantity": 6,
    },
  ],
  "total": Object {
    "amount": 550,
    "currency": "BRL",
    "precision": 2,
  },
}
`;
